### ISSUE TITLE
ChangeState example broken

### ISSUE THREAD

**Issue by adamgundry** [acid-state/acid-state#102]
With current master, the `ChangeState` error example does not throw an error for me (on GHC 8.0.2, with safecopy 0.9.4.1), even though it should. On the other hand if I add a call to `createCheckpoint secondAcid`, or to an update that tries to read the state, I get this:
```
This example simulates what happens when you modify your state type
without telling AcidState how to migrate from the old version to the new.
Hopefully this program will fail with a readable error message.

ChangeState.hs: Could not parse saved checkpoint due to the following error: too few bytes
From:	Main.SecondState:
	demandInput


CallStack (from HasCallStack):
  error, called at src/Data/Acid/Local.hs:398:7 in acid-state-0.16.0-inplace:Data.Acid.Local
ChangeState.hs: thread blocked indefinitely in an MVar operation
```
I guess that somehow `safecopy` isn't noticing the change of type, and merely opening and closing the component doesn't even attempt to read the checkpoint, but interacting with the component causes the checkpoint to be read back at the wrong type.

**Comment by stepcut** [comment]
I have some mild concerns about this. You seem to be saying that `openLocalStateFrom` will return successfully even if the underlying database is corrupt/mismatched. That makes me wonder:

 1) Was that always the case?

 2) Is that actually the behavior that we want?

Also, in your fix you have:

```
$(deriveSafeCopy 0 'base ''FirstState)
$(deriveSafeCopy 1 'base ''SecondState)
-- The version number difference is important here, as safecopy has no
-- way to notice if we change a type and fail to update its migration
-- history.
```

But I think that bumping the version number is missing the point of the test. I believe the point of the test is to show what would happen if someone comes and modifies the type but doesn't do anything at all with safecopy aspects. 

The idea is that at first the program has:

```
data MyState = MyState String
    deriving (Show)
$(deriveSafeCopy 0 'base ''MyState)
```

The program is run and creates a checkpoint. Then some developer comes along and decides that `Text` is better than `String` so they change the program to:

```
data MyState = MyState Text
    deriving (Show)
$(deriveSafeCopy 0 'base ''MyState)
```

If you change the type but don't update safecopy then you would expect the error message you get above,

```
ChangeState.hs: Could not parse saved checkpoint due to the following error: too few bytes
From:	Main.SecondState:
	demandInput
```

And that is because `String` is serialized as a list of Char but `Text` is serialized as a `ByteString`. So the parser *should* fail. 

The correct fix, of course, is to provide instances for the old and new state and provide a migration instance.

Obviously, having to run the program, change the code, and run it again is a rather tricky test case to implement, so the original test simulated that by having `FirstState` and `SecondState` and just doing two opens/closes in the same app.

I am unclear how `openLocalStateFrom` could succeed with the original test. Looking at the code it seems like it would have to have forced this value before it could return,

https://github.com/acid-state/acid-state/blob/master/src/Data/Acid/Local.hs#L302

And, looking at older versions of the code, I don't see a time when it is done differently enough that it would have somehow gotten lazier since this test was implemented. But apparently I am wrong about at least one of those things if not both.

In summary, I am wondering if perhaps the test was showing a bug, and instead of fixing the bug, we changed test to work around the bug.

**Comment by adamgundry** [comment]
Sorry, I think the bug report above is misleading... I now get the expected failure when testing with various GHC versions. I'm not sure which tree led to the bug above, but it may have been an interim version related to #96 rather than a build of master. Thus I think you are right and `openLocalStateFrom` is fine, but this is all the more reason to get this automatically tested properly lest it breaks in the future. :wink:

A problem I subsequently hit when automating the test (#108) was that changing the payload type of `SecondState` from `Text` to `Int` makes the error disappear, presumably because the checkpoint happens to continue to parse despite the change. And this made me question the logic of the test in the first place. If the state type is changed without any safecopy-related updates, this might lead to a parse failure (as in the case of `String` to `Text`), or it might not. AFAICS safecopy doesn't provide a way to catch this. That was my rationale for changing the test to something that guarantees safecopy will spot an error (the version mismatch).

I'm happy to restore the previous version of the test, if you think that's better. I just think we need to be clearer to distinguish the problems safecopy will reliably catch, from those where it may silently miss a change if the serialised checkpoint happens to continue to parse.

**Comment by stepcut** [comment]
Yes. It is unfortunate that sometimes the data will happen to parse, even though it is clearly wrong. I sometimes wonder if it would be possible to use TH/Generics/etc to create a 'fingerprint' that could be used to detect when someone changed the types but did not add proper migration instances.

It sounds like we maybe just need two tests -- since they are testing different things -- and are both things worth testing.

Ultimately, this is a reflection of one of my biggest complaints about acid-state+safecopy -- which is the lack of a schema in the data.

The work to allow different non-safecopy backends is great, because safecopy has some pretty serious design shortcomings that are not trivial to fix.

**Comment by adamgundry** [comment]
Okay, I've updated #108 accordingly.

> I sometimes wonder if it would be possible to use TH/Generics/etc to create a 'fingerprint' that could be used to detect when someone changed the types but did not add proper migration instances.

There's been some discussion related to this on [GHC Trac 7897](https://ghc.haskell.org/trac/ghc/ticket/7897). I guess it is in principle possible with TH/Generics though. Overall I think the approach of acid-state offering a choice of backends, perhaps with different design choices related to migration, is a good one.
